File: src/pratice/backtraking/연산자끼워넣기14888.java
package pratice.backtraking;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class 연산자끼워넣기14888 {
    public static BufferedReader br= new BufferedReader(new InputStreamReader(System.in));
    public static int[] num;
    public static int[] selectedOperator;
    public static int[] operatorCnt = new int[4];
    static int N;
    public static int max = Integer.MIN_VALUE;
    public static int min = Integer.MAX_VALUE;
    public static int cnt = 0;

    public static void main(String[] args) throws IOException {
        input();
        function(0);
        System.out.println(max);
        System.out.println(min);
    }

    private static void function(int curIdx) {
        if(curIdx >= N - 1){
            cnt++;
            int result = result();
            if(result > max) max = result;
            if(result < min) min = result;
        }else{
            for (int i = 0; i < 4; i++) {
                if(operatorCnt[i] > 0){
                    operatorCnt[i]--;
                    selectedOperator[curIdx] = i;
                    function(curIdx + 1);
                    operatorCnt[i]++;
                }
            }
        }
    }

    public static int result(){
        int result = num[0];
        for (int i = 0; i < selectedOperator.length; i++) {
            result = calculate(result, num[i + 1], selectedOperator[i]);
        }
        return result;
    }
    public static int calculate(int numA, int numB, int express){
        if(express == 0){
            return numA + numB;
        }else if(express == 1){
            return numA - numB;
        }else if(express == 2){
            return numA * numB;
        }else if(express == 3){
            return numA / numB;
        }else{
            throw new IllegalArgumentException("오류");
        }
    }

    private static void input() throws IOException {
        N = Integer.parseInt(br.readLine());
        String[] numStr = br.readLine().split(" ");
        num = new int[N];
        selectedOperator = new int[N - 1];
        for (int i = 0; i < N; i++) {
            num[i] = Integer.parseInt(numStr[i]);
        }
        String[] exCntStr = br.readLine().split(" ");
        for (int i = 0; i < 4; i++) {
            operatorCnt[i] = Integer.parseInt(exCntStr[i]);
        }

    }
}

File: src/pratice/backtraking/연산자끼워넣기14888.kt
package pratice.backtraking

//n - 1 개 카드로 중복 없이 n-1개를 순서있게 나열하는 방법

private var N = 0
private var num = intArrayOf()
private var chars = charArrayOf('+', '-', '*', '/')
private val selected = mutableListOf<Char>()
private var operator = listOf<Char>()
private var used = BooleanArray(0)
private var max = Int.MIN_VALUE
private var min = Int.MAX_VALUE

private fun main(){
    input()
    func(0)
    println(max)
    println(min)
}

private fun func(k: Int) {
    if(k >= N - 1){
        val result = cal()
        if(result > max) max = result
        if(result< min) min = result
    }else{
        for(i in 0 until N - 1){
            if(!used[i]){
                selected.add(operator[i])
                used[i] = true
                func(k + 1)
                used[i] = false
                selected.removeLast()
            }
        }
    }
}

fun cal(): Int {
    var sum = num[0]
    for((idx, c) in selected.withIndex()){
        when(c){
            '+' -> { sum += num[idx + 1]}
            '-' -> { sum -= num[idx + 1]}
            '/' -> {  sum = if(sum > 0)   sum / num[idx + 1] else -((-sum) / num[idx + 1])}
            '*' -> { sum *= num[idx + 1]}
        }
    }
    return sum
}

private fun input() {
    N = readln().toInt()
    num = readln().split(" ").map { it.toInt() }.toIntArray()
    operator = readln().split(" ").map { it.toInt() }.flatMapIndexed { idx, num -> List<Char>(num) { chars[idx] } }
    used = BooleanArray(N - 1)
}